### ISSUE TITLE
[Proposal] Automatically chain nodes placed on top of existing connection

### ISSUE THREAD

**Issue by fccDunce** [RodZill4/material-maker#156]
**Feature/enhancement description:**

Perform automatic chaining of nodes when dropped on top of connection wire.
---
Example- A Perlin Noise node's output is connected to a Warp node's input. Dragging a new Tones node over the connection wire highlights it, and if dropped the nodes are automatically rewired such that the output from the Perlin noise node is connected to the input of the Tones node, and the output of the Tones node is connected to the Warp node's input.

Two extra behaviors to complement it:
a) Deleting the middle node should keep the connections when possible. This may be the desired/expected behavior but can also be performed only when holding down a modifier key.

b) Right-clicking a connection wire should launch an "add node" command, and automatically place the new node at the middle of the existing connection.

(Similar behaviors already exists in several other node-based editors.)

**Comment by RodZill4** [comment]
That would be great, but I think none of those behaviors are supported in Godot's GraphEdit. These will definitely implemented, but I cannot guarantee this will be short term improvements.

**Comment by ikonnow** [comment]
What is the status here? I believe that even a small step in this direction could be absolutely groundbreaking. _Therefore, for example, the feature of Blender can be taken as a model._ 
If you observe the absolute basic interaction of the user with nodes, connecting new nodes is a constant repetition and costs unnecessary time and clicks.

**Comment by ikonnow** [comment]
> That would be great, but I think none of those behaviors are supported in Godot's GraphEdit. These will definitely implemented, but I cannot guarantee this will be short term improvements.

So there is no optimization or automatic connection of the nodesockets until today? I find that really hard to believe... 🤔

**Comment by RodZill4** [comment]
> > That would be great, but I think none of those behaviors are supported in Godot's GraphEdit. These will definitely implemented, but I cannot guarantee this will be short term improvements.
> 
> So there is no optimization or automatic connection of the nodesockets until today? I find that really hard to believe... 🤔

Connectiung programmatically is possible, but there's no way of detecting a node has been dragged into a connection (AFAIK).

**Comment by RodZill4** [comment]
> What is the status here? I believe that even a small step in this direction could be absolutely groundbreaking. _Therefore, for example, the feature of Blender can be taken as a model._ If you observe the absolute basic interaction of the user with nodes, connecting new nodes is a constant repetition and costs unnecessary time and clicks.

I generally connect into an empty space to create (and automatically connect) nodes. that's different from Blender but better than nothing.

**Comment by ikonnow** [comment]
> > What is the status here? I believe that even a small step in this direction could be absolutely groundbreaking. _Therefore, for example, the feature of Blender can be taken as a model._ If you observe the absolute basic interaction of the user with nodes, connecting new nodes is a constant repetition and costs unnecessary time and clicks.
> 
> I generally connect into an empty space to create (and automatically connect) nodes. that's different from Blender but better than nothing.

True, this is super good and is used almost "constantly" (probably not only by me), - in blender a small + is displayed for this, as soon as you pull a cable from a Nodesocket into the "empty". 
To briefly describe my observation of the current function: There is already a "highlighting" of the connection as soon as connected nodes are clicked/moved, which is why I assumed that the connection can be controlled individually and triggered by nodes that are moved over a connection (just like the logic in blender). 
Anyway, I'm curious, because in fact you do that all the time - you connect nodes and realize, oh, I need another node in between, then comes the "rewiring".

**Comment by williamchange** [comment]
It's implemented in Godot, so perhaps this could be a good reference

https://github.com/godotengine/godot/pull/83510

![linked_ref](https://github.com/godotengine/godot/assets/50084500/35f43565-6691-40b8-89ef-a6f894cb1720)
